File: gen_cnames.go
package main

import (
	"flag"
	"io/ioutil"
	"os"
	"regexp"
	"strings"

	"github.com/cbroglie/mustache"
)

var (
	maxDomains int = 16
	users      map[string][]string
	validate   = regexp.MustCompile(`[a-zA-Z0-9\-\.]+\.[a-zA-Z]+`)
)

func init() {
	flag.IntVar(&maxDomains, "max-domains", 16, "Max domains")
	flag.Parse()
}

func main() {
	dirs, _ := ioutil.ReadDir("/home/")
	for _, user := range dirs {
		cname := "/home/" + user.Name() + "/.CNAME"
		var domains []string
		if b, err := os.ReadFile(cname); err == nil {
			for _, v := range strings.Split(string(b), `\n`) {
				if validate.MatchString(v) {
					domains = append(domains, v)
				}
			}
			domains = domains[:(maxDomains - 1)]
			users[user.Name()] = domains
		}
	}

	http, _ := mustache.RenderFile("nginx.mustache", users)
	_ = ioutil.WriteFile("/tmp/result5.http", []byte(http), 0644)

	https, _ := mustache.RenderFile("nginx.ssl.mustache", users)
	_ = ioutil.WriteFile("/tmp/result5.https", []byte(https), 0644)
}

File: gen_cnames.py
import os
import re
import chevron

MAX_DOMAINS = 16

nginx_template = ""
nginx_ssl_template = ""
validate = re.compile("[a-zA-Z0-9\-\.]+\.[a-zA-Z]+").match

with open("nginx.mustache") as fp:
    nginx_template = fp.read()

with open("nginx.ssl.mustache") as fp:
    nginx_ssl_template = fp.read()

users = list()

for username in os.listdir("/home/"):
    cname = f"/home/{username}/.CNAME"
    if os.path.isfile(cname):
        with open(cname) as fp:
            domains = map(
                lambda domain, _: domain.strip() if validate(domain) else "",
                fp.readlines(),
                range(MAX_DOMAINS)
            )
            domains = tuple(domains)
            users.append(
                dict(username=username, domains=domains)
            )

with open("/tmp/result5.http", "w") as fp:
    fp.write(chevron.render(nginx_template, dict(users=users)))

with open("/tmp/result5.https", "w") as fp:
    fp.write(chevron.render(nginx_ssl_template, dict(users=users)))
